File: mbw/src/androidTest/java/com/mycelium/wallet/activity/ViewVisibilityIdlingResource.java
package com.mycelium.wallet.activity;

import android.app.Activity;
import android.os.Handler;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.test.espresso.*;
import android.view.View;

import java.lang.ref.WeakReference;

public class ViewVisibilityIdlingResource implements IdlingResource {

    private static final int IDLE_POLL_DELAY_MILLIS = 100;

    /** Hold weak reference to the View, so we don't leak memory even if the resource isn't unregistered. */
    private final WeakReference<View> mView;
    private final int mVisibility;
    private final String mName;

    private ResourceCallback mResourceCallback;

    /**
     * @param activity which owns the View
     * @param viewId ID of the View to monitor
     * @param visibility One of {@link View#VISIBLE}, {@link View#INVISIBLE}, or {@link View#GONE}.
     */
    public ViewVisibilityIdlingResource(@NonNull Activity activity, @IdRes int viewId, int visibility) {
        this(activity.findViewById(viewId), visibility);
    }

    /**
     * @param view the View to monitor
     * @param visibility One of {@link View#VISIBLE}, {@link View#INVISIBLE}, or {@link View#GONE}.
     */
    public ViewVisibilityIdlingResource(@NonNull View view, int visibility) {
        mView = new WeakReference<View>(view);
        mVisibility = visibility;
        mName = "View Visibility for view " + view.getId() + "(@" + System.identityHashCode(mView) + ")";
    }

    @Override
    public String getName() {
        return mName;
    }

    @Override
    public boolean isIdleNow() {
        View view = mView.get();
        final boolean isIdle = view == null || view.getVisibility() == mVisibility;
        if (isIdle) {
            if (mResourceCallback != null) {
                mResourceCallback.onTransitionToIdle();
            }
        } else {
            /* Force a re-check of the idle state in a little while.
             * If isIdleNow() returns false, Espresso only polls it every few seconds which can slow down our tests.
             * Ideally we would watch for the visibility state changing, but AFAIK we can't detect when a View's
             * visibility changes to GONE.
             */
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    isIdleNow();
                }
            }, IDLE_POLL_DELAY_MILLIS);
        }

        return isIdle;
    }

    @Override
    public void registerIdleTransitionCallback(ResourceCallback resourceCallback) {
        mResourceCallback = resourceCallback;
    }
}

File: mbw/src/androidTest/java/com/mycelium/wallet/activity/ViewVisibilityIdlingResource.kt
package com.mycelium.wallet.activity

import android.app.Activity
import android.os.Handler
import android.view.View

import androidx.annotation.IdRes
import androidx.test.espresso.IdlingResource

import java.lang.ref.WeakReference

class ViewVisibilityIdlingResource(view: View, private val visibility: Int) : IdlingResource {
    private val viewWeakReference: WeakReference<View> = WeakReference(view)
    private val name: String

    private var resourceCallback: IdlingResource.ResourceCallback? = null

    constructor(activity: Activity, @IdRes viewId: Int, visibility: Int) : this(activity.findViewById<View>(viewId), visibility)

    init {
        name = "View Visibility for view " + view.id + "(@" + System.identityHashCode(viewWeakReference) + ")"
    }

    override fun getName(): String = name

    override fun isIdleNow(): Boolean {
        val view = viewWeakReference.get()
        val isIdle = view == null || view.visibility == visibility
        if (isIdle) {
            if (resourceCallback != null) {
                resourceCallback!!.onTransitionToIdle()
            }
        } else {
            Handler().postDelayed({ isIdleNow }, IDLE_POLL_DELAY_MILLIS)
        }

        return isIdle
    }

    override fun registerIdleTransitionCallback(resourceCallback: IdlingResource.ResourceCallback) {
        this.resourceCallback = resourceCallback
    }

    companion object {
        private const val IDLE_POLL_DELAY_MILLIS = 100L
    }
}